scipy/sparse/_sputils.py

```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
         new_shape = tuple(operator.index(arg) for arg in args)
 
     if current_shape is None:
-        if allow_1d:
-            if len(new_shape) not in (1, 2):
-                raise ValueError('shape must be a 1- or 2-tuple of positive '
-                                 'integers')
-        elif len(new_shape) != 2:
-            raise ValueError('shape must be a 2-tuple of positive integers')
-        if any(d < 0 for d in new_shape):
-            raise ValueError("'shape' elements cannot be negative")
+        if len(new_shape) < 3:
+            if allow_1d:
+                if len(new_shape) not in (1, 2):
+                    raise ValueError('shape must be a 1- or 2-tuple of positive '
+                                     'integers')
+            elif len(new_shape) != 2:
+                raise ValueError('shape must be a 2-tuple of positive integers')
+            if any(d < 0 for d in new_shape):
+                raise ValueError("'shape' elements cannot be negative")
+        else: # dim >= 3
```

Comment:
This block is largely duplicated as well, can you try to reduce that (similar to my comment for `isshape`)?